### ISSUE TITLE
appcfg is deprecated: Correct command in GettingStarted Step 6

### ISSUE THREAD

**Issue by ryosuketc** [google/personfinder#771]
`tools/gae update app -A <project_id> --noauth_local_webserver` fails due to deprecation of `appcfg`.
Step 6 in [GettingStarted](https://github.com/google/personfinder/wiki/GettingStarted)

The command has to be migrated to `gcloud`.

```
 % tools/gae update app -A <project_id> --noauth_local_webserver
2020-11-25 00:19:36,626 WARNING appcfg.py:5217 The appcfg command is deprecated. Please download the Cloud SDK for
App Engine development. https://cloud.google.com/appengine/downloads

...(omitted)...
 
12:21 AM Application deployment failed. Message: Deployments using appcfg are no longer supported. See https://cloud.google.com/appengine/docs/deprecations
2020-11-25 00:21:44,800 ERROR appcfg.py:2564 Application deployment failed. Message: Deployments using appcfg are no longer supported. See https://cloud.google.com/appengine/docs/deprecations 
12:21 AM Rolling back the update.
Could not start serving the given version.
```

**Comment by ryosuketc** [comment]
DEPRECATED: 
```
tools/gae update app -A my-app-id --noauth_local_webserver
```
-> It eventually invokes:
https://github.com/google/personfinder/blob/8c31f8e6044b57e77e3898d8aaf32ed54bfdd3c8/tools/gae#L42

NEW ((my-app-id == Cloud project ID)):
```
gcloud app deploy app/app.yaml --project <my-app-id>
```

However, `application` and `version` needed to be removed to make it compatible with `gcloud` command.
See [app.yaml Reference  |  App Engine standard environment for Python 2](https://cloud.google.com/appengine/docs/standard/python/config/appref) as well as the following command line errors.

```
ERROR: (gcloud.app.deploy) The [application] field is specified in file [app.yaml file path]. This field is not used by gcloud and must be removed. Project name should instead be specified either by `gcloud config set project MY_PROJECT` or by setting the `--project` flag on individual command executions.

ERROR: (gcloud.app.deploy) The [version] field is specified in file [app.yaml file path]. This field is not used by gcloud and must be removed. Versions are generated automatically by default but can also be manually specified by setting the `--version` flag on individual command executions.
```

Note: While executing the NEW command, you might need to grant `Storage Object Viewer` permission to a service account. Follow error messages.

1.   Can we just remove `application` and `version` from `app.yaml`?
    -   `version` is supposed to start from a letter instead of a number as per the app.yaml document above.
1.  `--noauth_local_webserver` equivalent in `gcloud`

For the first bullet, since `appcfg` is already deprecated and not work, we have to fix any errors as they occur (assuming `app.yaml` is only used by the command). We could remove the lines anyway and see if it causes any errors in the future.

**Comment by gimite** [comment]
Thanks for taking a look!

> For the first bullet, since appcfg is already deprecated and not work, we have to fix any errors as they occur (assuming app.yaml is only used by the command)

Makes sense. Can you proceed with the change to remove `application` and `version`? (The server test might have dependency to this (I don't know), but I guess we should update the server test as well.)

> --noauth_local_webserver equivalent in gcloud

Not really sure if this is necessary. Can you try without it first, and if it doesn't cause any issues, I guess we won't need it.

**Comment by ryosuketc** [comment]
The approach makes sense. Let me see if what we need to modify after the change

**Comment by ryosuketc** [comment]
Looking at an internal documentation on releasing PF, the existing `appcfg` command expects an original version continue running until we migrate traffic manually from console (or CLI or whatever).
Hence the deploy command should **not** [promote](https://cloud.google.com/sdk/gcloud/reference/app/deploy#--promote) the deployed version to receive all traffic.
```
gcloud app deploy app/app.yaml --project rtachibana-dev --no-promote
```

**Comment by ryosuketc** [comment]
The command succeeds (i.e. deploy the app) when 1. it is directly invoked from CLI OR 2. it is invoked from `tools/gae` with `source common.sh` being commented out.
It fails, however when it is invoked from `tools/gae` with `source/common.sh` as it is.

@gimite @myoshimu I am investigating the error to migrate from `appcfg` to `gcloud`. I am suspecting any variable set by `common.sh` might be unexpectedly affecting `gcloud`. Could you let me know if you come up with any clue?

**1. SUCCESS:**
Directly invoke the command from shell.
```
$ gcloud app deploy app/app.yaml --project rtachibana-dev --no-promote
```

**2. SUCCESS:**
1.  Change this line
https://github.com/google/personfinder/blob/8c31f8e6044b57e77e3898d8aaf32ed54bfdd3c8/tools/gae#L42
to:
```
gcloud "$@"
```
2. Comment out this line
https://github.com/google/personfinder/blob/8c31f8e6044b57e77e3898d8aaf32ed54bfdd3c8/tools/gae#L17

3.  Run
```
$ tools/gae app deploy app/app.yaml --project rtachibana-dev --no-promote
```

**3. FAIL:**
Try the second case without commenting out `source common.sh` line.

```
ERROR: gcloud failed to load (gcloud.app.deploy): Problem loading gcloud.app.deploy: No module named api.

This usually indicates corruption in your gcloud installation or problems with your Python interpreter.

Please verify that the following is the path to a working Python 2.7 or 3.5+ executable:
    /usr/local/google/home/rtachibana/repos/pf/venv/bin/python
If it is not, please set the CLOUDSDK_PYTHON environment variable to point to a working Python 2.7 or 3.5+ executable.

If you are still experiencing problems, please run the following command to reinstall:
    $ gcloud components reinstall

If that command fails, please reinstall the Cloud SDK using the instructions here:
    https://cloud.google.com/sdk/
```

**Comment by gimite** [comment]
Looks like `tools/gae app ...` is equivalent of directly hitting `appcfg.py app ...`:
https://github.com/google/personfinder/blob/master/tools/gae#L42

How about just hitting your gcloud command instead of running `tools/gae app ...`? Maybe it's the easiest solution for this particular case?

**Comment by ryosuketc** [comment]
Indeed, the difference is whether we run `common.sh` or not. If we do not run it, the `gcloud` command works fine.
But my concern is whether something go wrong because of the lack of 'common.sh', which sets some environment variables and is supposed to be sourced in every files under `tools/` dir.

**Comment by gimite** [comment]
Just to clarify, my suggestion is to let developers (and the release script) directly run:

```
gcloud app deploy app/app.yaml --project rtachibana-dev --no-promote
```

instead of using `tools/gae`. In this scenario, I believe not running `common.sh` should cause no issues because `gcloud` command should have no dependency to any of the environment variables defined there. WDYT?

**Comment by ryosuketc** [comment]
tl;dr is I agree with you to have developers run `gcloud` commands directly where no wrapper scripts are needed. Once I get your confirmation, I will update GettingStarted and close this issue.

**Details**
I pinpointed the cause - `APPENGINE_RUNTIME=python27` set by `common.sh` somehow prevents `gcloud`. Indeed, when I unset it, the command works fine.
-   The variable seems to determine the runtime ([source](https://stackoverflow.com/questions/29783483/how-can-i-tell-the-code-is-running-on-google-app-engine)) but I can no longer find it in the [official reference](https://cloud.google.com/appengine/docs/standard/python/runtime#environment_variables_2). Probably it is no longer used in the recent GAE environments.
- It is used in `urllib3` library in `app/vendors`.

Possible solutions are the following.  I would prefer 1 for this specific command (`tools/gae update` since `tools/gae` add no value to it) and continue investigating broader use cases such as `tools/gae run' in https://github.com/google/personfinder/issues/395.

1.  Directly run `gcloud`.
2. `unset APPENGINE_RUNTIME` temporarily right before `gcloud` and `source common.sh` right after it.
3. Remove this line but it causes failure in unit tests. https://github.com/google/personfinder/blob/d845d5a801565a512e4501bb34c03f3ca789a125/tools/common.sh#L90

```
 % tools/all_tests
No config file found, using default configuration

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)


--- Running unit tests

===================================================================================== ERRORS ======================================================================================
_______________________________________________________________________ ERROR collecting tests/test_api.py ________________________________________________________________________
../tests/test_api.py:22: in <module>
    import api
api.py:42: in <module>
    import cloud_storage
cloud_storage.py:26: in <module>
    import config
config.py:26: in <module>
    import utils
utils.py:43: in <module>
    import google.appengine.ext.webapp.template
../../../opt/google_appengine/google/appengine/ext/webapp/template.py:63: in <module>
    webapp._config_handle.django_setup()
../../../opt/google_appengine/google/appengine/ext/webapp/__init__.py:116: in _django_setup
    getattr(django.conf.settings, 'FAKE_ATTR', None)
vendors/django/conf/__init__.py:56: in __getattr__
    self._setup(name)
vendors/django/conf/__init__.py:39: in _setup
    % (desc, ENVIRONMENT_VARIABLE))
E   ImproperlyConfigured: Requested setting FAKE_ATTR, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
______________________________________________________________________ ERROR collecting tests/test_create.py ______________________________________________________________________
../tests/test_create.py:25: in <module>
    import create
create.py:16: in <module>
    import photo
photo.py:20: in <module>
    import utils
utils.py:43: in <module>
    import google.appengine.ext.webapp.template
../../../opt/google_appengine/google/appengine/ext/webapp/template.py:63: in <module>
    webapp._config_handle.django_setup()
../../../opt/google_appengine/google/appengine/ext/webapp/__init__.py:116: in _django_setup
    getattr(django.conf.settings, 'FAKE_ATTR', None)
vendors/django/conf/__init__.py:56: in __getattr__
    self._setup(name)
vendors/django/conf/__init__.py:39: in _setup
    % (desc, ENVIRONMENT_VARIABLE))
E   ImproperlyConfigured: Requested setting FAKE_ATTR, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
_________________________________________________________________ ERROR collecting tests/test_full_text_search.py _________________________________________________________________
../tests/test_full_text_search.py:26: in <module>
    import delete
delete.py:15: in <module>
    import reveal
reveal.py:28: in <module>
    from recaptcha.client import captcha
recaptcha/client/captcha.py:7: in <module>
    import config
config.py:26: in <module>
    import utils
utils.py:43: in <module>
    import google.appengine.ext.webapp.template
../../../opt/google_appengine/google/appengine/ext/webapp/template.py:63: in <module>
    webapp._config_handle.django_setup()
../../../opt/google_appengine/google/appengine/ext/webapp/__init__.py:116: in _django_setup
    getattr(django.conf.settings, 'FAKE_ATTR', None)
vendors/django/conf/__init__.py:56: in __getattr__
    self._setup(name)
vendors/django/conf/__init__.py:39: in _setup
    % (desc, ENVIRONMENT_VARIABLE))
E   ImproperlyConfigured: Requested setting FAKE_ATTR, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
3 error in 1.71 seconds
```

**Comment by gimite** [comment]
I prefer the option 1. Thanks!

**Comment by myoshimu** [comment]
Hi Ryosuke san,
When I simply tried `gcloud app deploy` it worked fine.
Did you set up the APPENGINE_DIR to the app engine directory as below?
(e.g.) I installed GCP SDK under /opt
APPENGINE_DIR=/opt/google-cloud-sdk/platform/google_appengine

**Comment by ryosuketc** [comment]
@myoshimu Thank you for pointing it out.
I was not aware of the `platform/google_appengine` dir in Cloud SDK. Considering `appcfg` is deprecated and standalone App Engine SDK is no longer available for download ([official doc](https://cloud.google.com/appengine/docs/standard/python/tools/appcfg-arguments)), I would guess the dir is a kind of relief measures for those who still rely on `appcfg` (but no longer recommended to be used).

I will go ahead and migrate out of `appcfg` as agreed.

**Comment by gimite** [comment]
Mentioned in the meeting, but APPENGINE_DIR contains two things in my understanding:

* Command line tools like `appcfg`
* AppEngine Specific libraries for Python (like datastore)

The former (or at least `appcfg` specifically) looks deprecated while I guess the latter isn't. @myoshimu will confirm it (Thanks!), but I guess we can continue depending on it so far.

**Comment by ryosuketc** [comment]
@myoshimu As mentioned in the meeting too but it would be great if we could know `--noauth_local_webserver` equivalent for `gcloud`.

**Comment by gimite** [comment]
As I mentioned somewhere above, I personally don't care much about `--noauth_local_webserver`. I don't quite remember what it is for, but it should be fine if you can run it without the flag with no issue.

**Comment by ryosuketc** [comment]
Here are some updates from @myoshimu 

> Mentioned in the meeting, but APPENGINE_DIR contains two things in my understanding:
> 
> * Command line tools like `appcfg`
> * AppEngine Specific libraries for Python (like datastore)

No plan to deprecate libraries under `APPENGINE_DIR=google-cloud-sdk/platform/google_appengine`

> 
> The former (or at least `appcfg` specifically) looks deprecated while I guess the latter isn't. @myoshimu will confirm it (Thanks!), but I guess we can continue depending on it so far.

> As I mentioned somewhere above, I personally don't care much about `--noauth_local_webserver`. I don't quite remember what it is for, but it should be fine if you can run it without the flag with no issue.

--no-launch-browser
https://cloud.google.com/sdk/gcloud/reference/auth/login?hl=ja#--launch-browser

**Comment by ryosuketc** [comment]
Refer to https://github.com/google/personfinder/pull/773#issuecomment-774863795 for the summary of the fix of the issue.
Based on the [search results](https://github.com/google/personfinder/search?q=repo%3Agoogle%2Fpersonfinder+tools%2Fgae&type=wikis), some documents are updated accordingly.

Compare Revisions · google/personfinder Wiki
https://github.com/google/personfinder/wiki/GettingStarted/_compare/a81c4d55431fe8649ba6e63feb233109c1163313...0ccecc3df83a2336ad86c749b7d4336072d39cba
Compare Revisions · google/personfinder Wiki
https://github.com/google/personfinder/wiki/Internationalization/_compare/f2f36615910637c9807eda3d49f05f4140feb269...cce27a2148bd9c9d6f28f6e23576cbe795e995a3
Compare Revisions · google/personfinder Wiki
https://github.com/google/personfinder/wiki/Launching/_compare/f2f36615910637c9807eda3d49f05f4140feb269...977245517a4446f302ad1e4ab096c2c66cc2b7e5
Compare Revisions · google/personfinder Wiki
https://github.com/google/personfinder/wiki/DeveloperGuideGHC11/_compare/f2f36615910637c9807eda3d49f05f4140feb269...d80204d85ee3117217b284ace1bf3c3be091dd6b
Compare Revisions · google/personfinder Wiki
https://github.com/google/personfinder/wiki/DeveloperFaq/_compare/0c0a4ba50a581538a70c87288601fccebd6866fe...637a4c59e17aaa0693c33fb5422986f200645eb2
